flowmachine/flowmachine/features/subscriber/meaningful_locations.py

```diff
@@ -0,0 +1,259 @@
+    a clustering of the cell towers they use, and their usage patterns for those towers.
+
+    This is an extension of work by Isaacman et al.[1]_ by Flowminder in collaboration with the World Bank[2]_.
+
+    Parameters
+    ----------
+    clusters : HartiganCluster
+        Per subscriber clusters of towers
+    scores : EventScore
+        Per user, per tower scores based on hour of day and day of week of interactions with the tower
+    labels : LabelEventScore
+        Labels to apply to clusters given their usage pattern scoring
+    label : str
+        Meaningful label to extract clusters for
+
+    References
+    ----------
+    .. [1] S. Isaacman et al., "Identifying Important Places in People's Lives from Cellular Network Data", International Conference on Pervasive Computing (2011), pp 133-151.
+    .. [2] Zagatti, Guilherme Augusto, et al. "A trip to work: Estimation of origin and destination of commuting patterns in the main metropolitan regions of Haiti using CDR." Development Engineering 3 (2018): 133-165.
+    """
+
+    def __init__(
+        self,
+        *,
+        clusters: HartiganCluster,
+        scores: EventScore,
+        labels: Dict[str, Dict[str, Any]],
+        label: str,
+    ) -> None:
+        labelled_clusters = LabelEventScore(
+            scores=clusters.join_to_cluster_components(scores), labels=labels
+        )
+        self.labelled_subset = labelled_clusters.subset("label", label)
+
+        super().__init__()
+
+    @property
+    def column_names(self) -> List[str]:
+        return ["subscriber", "label", "cluster", "n_clusters"]
+
+    def _make_query(self):
+        return f"""
+        SELECT subscriber, label, cluster, (sum(1) OVER (PARTITION BY subscriber)) as n_clusters FROM 
+            ({self.labelled_subset.get_query()}) clus 
+        GROUP BY subscriber, label, cluster
+        ORDER BY subscriber
+        """
+
+
+class MeaningfulLocationsAggregate(Query):
+    """
+    Aggregates an individual level meaningful location to a spatial unit by assigning
+    subscribers with clusters in that unit to it. For subscribers with more than one cluster,
+    assigns `1/n_clusters` to each spatial unit that the cluster lies in.
+
+    Parameters
+    ----------
+    meaningful_locations : MeaningfulLocations
+        A per-subscriber meaningful locations object to aggregate
+    level : {"admin3", "admin2", "admin1", "grid", "polygon"}, default "admin3"
+        Spatial unit to aggregate to
+    column_name : str or list of str, default None
+        Optionally specify a non-default column name or names from the spatial unit table
+    polygon_table : str, default None
+        When using the "polygon" level, you must specify the fully qualified name of a table
+        containing polygons.
+    geom_column : str, default "geom"
+        When using the "polygon" level, you must specify the name of column containing geometry
```

Comment:
Substitute "level" with "aggregation unit"?

Reply:
Gonna skip that for now, because all of the test harness stuff assumes level is the name of the arg.

Reply:
Ok 👍